### ISSUE TITLE
Multiple possible request types for an endpoint

### ISSUE THREAD

**Issue by harpocrates** [julienrf/endpoints#529]
It would be useful to allow multiple different types of request types. For instance `jsonRequest[Foo].orElse(textRequest): RequestEntity[Either[Foo, String]]`. However, I'm not sure how to statically rule out cases where the two request types are actually the same (ex: `jsonRequest[Foo].orElse(jsonRequest[Bar])`). Maybe it would suffice to document this outside the type system?

I know OpenAPI supports documenting multiple request types for an endpoint.

**Comment by julienrf** [comment]
That makes sense.

> However, I'm not sure how to statically rule out cases where the two request types are actually the same (ex: `jsonRequest[Foo].orElse(jsonRequest[Bar])`)

In that specific case, maybe `jsonRequest(fooSchema orElse barSchema)`?

**Comment by harpocrates** [comment]
Any chance this could make it into 1.0.0? @ethanb29 has an implementation with a couple of the interpreters already done.

**Comment by julienrf** [comment]
Yes.

**Comment by julienrf** [comment]
Hey @harpocrates, any update on this?

**Comment by harpocrates** [comment]
I think #544 is the only difficult part. After that, this should be quick straightforward to implement.

**Comment by harpocrates** [comment]
I may be missing something, but I think this will be impossible to implement for Play's `BodyParser` (there doesn't seem to be a way to disentangle the parsing from the validation). I've got all the other cases implemented though. Assuming this really is impossible for Play, I think that leaves two options:

  * make some trait extending `Requests` and implement it in all clients/servers `Endpoints` implementation, except for Play
  
    ```scala
    trait EitherRequestEntities extends Requests {
      def choiceRequestEntity[A, B](
        requestEntityA: RequestEntity[A],
        requestEntityB: RequestEntity[B]
      ): RequestEntity[Either[A, B]]

      implicit class RequestEntitySyntax[A](requestEntity: RequestEntity[A]) {
        def orElse[B](otherRequestEntity: RequestEntity[B]): RequestEntity[Either[A, B]] =
          choiceRequestEntity(requestEntity, otherRequestEntity)
      }
    }
    ```

  * wontfix this issue

@julienrf are you OK with the first option? If so, do you have opinions for the names?

**Comment by julienrf** [comment]
For the Play server interpreter, we will probably have to change how the type `RequestEntity[A]` is fixed. It can’t anymore be just `BodyParser[A]`, but needs to be something like `RequestHeader => Option[BodyParser[A]]` (I think), so that we can implement `choiceRequestEntity` as follows:

~~~ scala
headers =>
  requestEntityA(headers).map(_.map(Left(_)))
    .orElse(requestEntityB(headers).map(_.map(Right(_)))
~~~

And `textRequest` should be adapted like so:

~~~ scala
lazy val textRequest: RequestEntity[String] =
  headers =>
    if (headers.contentType.exists(_.equalIgnoreCase("text/plain")) Some(playComponents.playBodyParsers.text)
    else None
~~~

In case I’m wrong or you don’t have time to investigate further this idea:

> are you OK with the first option?

Yes.

> do you have opinions for the names?

What do you think of `RequestEntityChoices`?

**Comment by harpocrates** [comment]
> For the Play server interpreter, we will probably have to change how the type RequestEntity[A] is fixed. It can’t anymore be just BodyParser[A], but needs to be something like RequestHeader => Option[BodyParser[A]] (I think), ...

I think I see what you have in mind: move the content type validation up into _endpoints_. I'll try this out and report back. If that doesn't work, `RequestEntityChoices` sounds good.

Thanks for your input.

**Comment by julienrf** [comment]
Hello! This is issue is the last one to address before we can cut the next release, is there anything we could do to help? I’m happy to help with the Play interpreters, for instance :)

**Comment by harpocrates** [comment]
I'm sorry - I've been thinking about this pretty much every day for the last week. Feel free to bump this off to the next version of endpoints if it is blocking a release 😓 (especially if that is an obstacle to Scala.js 1.0 support). If you have spare time to help, here is my WIP: https://github.com/harpocrates/endpoints/commit/ddc523fbc5f48f6dce184ce065d40184f4e7d4f8.

I've finished (but not yet tested) all interpreters except for the Play server. The challenge there is trying to produce good error messages (the `RequestHeader => Option[BodyParser[A]]` suggestion needs some tweaking to produce better error messages) without re-implementing too much logic that belongs in web-servers.

**Comment by julienrf** [comment]
Thank you @harpocrates! Here is my proposal for the play-server interpreter: https://github.com/julienrf/endpoints/commit/994355cb57cdb96a429f05b15b7ffcd93d62e691 Let me know what you think. There is room for cleaning things up, but I would like to do that later.

**Comment by harpocrates** [comment]
That looks good to me. I'll roll that into a PR this weekend. Thank you for putting it together. 😄
